clippy_dev/src/crater.rs

```diff
@@ -0,0 +1,182 @@
+        let extract_dir = PathBuf::from("target/crater/crates");
+        let krate_download_dir = PathBuf::from("target/crater/downloads");
+
+        // url to download the crate from crates.io
+        let url = format!(
+            "https://crates.io/api/v1/crates/{}/{}/download",
+            self.name, self.version
+        );
+        println!("Downloading and extracting {} {} from {}", self.name, self.version, url);
+        let _ = std::fs::create_dir("target/crater/");
+        let _ = std::fs::create_dir(&krate_download_dir);
+        let _ = std::fs::create_dir(&extract_dir);
+
+        let krate_file_path = krate_download_dir.join(format!("{}-{}.crate.tar.gz", &self.name, &self.version));
+        // don't download/extract if we already have done so
+        if !krate_file_path.is_file() {
+            // create a file path to download and write the crate data into
+            let mut krate_dest = std::fs::File::create(&krate_file_path).unwrap();
+            let mut krate_req = ureq::get(&url).call().unwrap().into_reader();
+            // copy the crate into the file
+            std::io::copy(&mut krate_req, &mut krate_dest).unwrap();
+
+            // unzip the tarball
+            let ungz_tar = flate2::read::GzDecoder::new(std::fs::File::open(&krate_file_path).unwrap());
+            // extract the tar archive
+            let mut archiv = tar::Archive::new(ungz_tar);
+            archiv.unpack(&extract_dir).expect("Failed to extract!");
+
+            // unzip the tarball
+            let ungz_tar = flate2::read::GzDecoder::new(std::fs::File::open(&krate_file_path).unwrap());
+            // extract the tar archive
+            let mut archiv = tar::Archive::new(ungz_tar);
+            archiv.unpack(&extract_dir).expect("Failed to extract!");
+        }
+        // crate is extracted, return a new Krate object which contains the path to the extracted
+        // sources that clippy can check
+        Krate {
+            version: self.version.clone(),
+            name: self.name.clone(),
+            path: extract_dir.join(format!("{}-{}/", self.name, self.version)),
+        }
+    }
+}
+
+impl Krate {
+    fn run_clippy_lints(&self, cargo_clippy_path: &PathBuf) -> Vec<String> {
+        println!("Linting {} {}...", &self.name, &self.version);
+        let cargo_clippy_path = std::fs::canonicalize(cargo_clippy_path).unwrap();
+
+        let shared_target_dir = clippy_project_root().join("target/crater/shared_target_dir/");
+
+        let all_output = std::process::Command::new(cargo_clippy_path)
+            .env("CARGO_TARGET_DIR", shared_target_dir)
+            // lint warnings will look like this:
+            // src/cargo/ops/cargo_compile.rs:127:35: warning: usage of `FromIterator::from_iter`
+            .args(&[
+                "--",
+                "--message-format=short",
+                "--",
+                "--cap-lints=warn",
+                "-Wclippy::pedantic",
```

Comment:
It would also be nice if we can specify the lint groups with a flag. I think this default is fine though.

Reply:
Hm, why not just lint everything?
That way we have greater coverage.

Reply:
Yeah, by default that is good, but for future testing, we might want to only run exactly one lint, to make it easier to check the output.